Experimental Features (#91)
buildpacks/rfcs: text/0000-experimental-features.md

We need a way to introduce experiments before writing an RFC. Such a mechanism would allow us to deliver new capabilities faster, improve our means of collecting real end-user feedback, and improve the quality of our RFCs.

# What it is
[what-it-is]: #what-it-is

- *experimental feature* - any behavior or options that are likely to change before being added to the specification

This proposal affects buildpack users, platform implementors, and project contributors who want to introduce experimental features. Today, any significant change that impacts end-users is introduced through an RFC, which requires a successful vote by the core team before being introduced into the specification (or the Pack platform).

We seek to change the current process such that experimental features do not require an RFC or specification. Instead, an experimental feature would only require approval by a simple majority of the relevant sub-teams maintainers.

# How it Works
[how-it-works]: #how-it-works

> Experimental features can be submitted as a PR or discussed in Github Issues on the relevant project. The team that maintains that project can decide whether or not to introduce the feature as long as it is behind a  mechanism that enables or disables these features.

Reviewer: I would also add that this mechanism needs to make it clear this is experimental, can change, etc.

# Drawbacks
[drawbacks]: #drawbacks

- End users may become tied to features that eventually go away or significantly change

# Alternatives
[alternatives]: #alternatives

## Plugins

A plugin system for Pack and/or Lifecycle would allow experimental features to be implemented completely outside of the CNB project. For example, a plugin could be installed to Pack with a command like:

```
$ pack install-plugin skele-gro
```

Then, a user can execute new commands provided by the plugin:

```
$ pack grow-bones
```

Or use new flags on existing commands:

```
$ pack build --with-bone-growth
```

> Such a plugin mechanism would require a well defined plugin-interface, and versioning of that interface. This is a significantly larger project that the current proposal. However, the current proposal does not prevent a plugin system from being introduced later on.

Reviewer 2: This is definitely something to consider.  Especially the work to maintain the integrity of the interface and how much effort we're willing to put into make sure we don't break the interface and any plugins.

Reviewer: I would mark plugins as out of scope.

- Start Date: 2020-07-02
- Author(s): [@jkutner](https://github.com/jkutner)
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: N/A

# Summary
[summary]: #summary

This is a proposal for allowing experimental features to enter Pack and Lifecycle without requiring a specification, core team review, or RFC.

# Motivation
[motivation]: #motivation

End-users and early adopters of CNB have real problems that need to be solved by the CNB project. Sometimes these problems prevent them from adopting CNB in their organizations. Our RFC process, specification guidelines, and governance structure provide a strong mechanism through which we can enact change and solve problems; but sometimes it's too heavy and inhibits experimentation.

We need a way to introduce experiments before writing an RFC. Such a mechanism would allow us to deliver new capabilities faster, improve our means of collecting real end-user feedback, and improve the quality of our RFCs.

# What it is
[what-it-is]: #what-it-is

- *experimental feature* - any behavior or options that are likely to change before being added to the specification

> This proposal affects buildpack users, platform implementors, and project contributors who want to introduce experimental features into Pack or Lifecycle. Today, any significant change that impacts end-users is introduced through an RFC, which requires a successful vote by the core team before being introduced into the specification (or the Pack platform).

Reviewer 3: I wonder if this particular mechanism for enabling experimental features is better suited to `pack` than the lifecycle. Not all lifecycle features can or should be enabled with flags (think buildpack API features).

I am wondering if experimental API versions as proposed in https://github.com/buildpacks/rfcs/pull/92 would better support this goal at the lifecycle level

> We seek to change the current process such that experimental features do not require an RFC or specification. Instead, an experimental feature would only require approval by a simple majority of Platform Maintainers.

Reviewer 4: ```suggestion
We seek to change the current process such that experimental features do not require an RFC or specification. Instead, an experimental feature would only require approval by a simple majority of the relevant sub-teams maintainers.
```

# How it Works
[how-it-works]: #how-it-works

> Experimental features can be submitted as a PR or discussed in Github Issues on the relevant project. The team that maintains that project can decide whether or not to introduce the feature as long as it is behind an `--experimental` flag.

Reviewer 5: Is this intended to mention our existing mechanism for gating experimental features? I know we've referenced it as an "experimental flag" in conversation but it's currently implemented in `pack` as a config property.

If so I would propose updating the wording to:

```suggestion
Experimental features can be submitted as a PR or discussed in Github Issues on the relevant project. The team that maintains that project can decide whether or not to introduce the feature as long as it is behind a  mechanism that enables or disables these features.
```

# Drawbacks
[drawbacks]: #drawbacks

- End users may become tied to features that eventually go away or significantly change

> 

Reviewer: The big concern I have with this is creating a set of "hidden" APIs that people depend on if we do decide to take it away.

Reviewer 2: I agree.  This seems like an unsustainable path of where we're having to constantly consider the ramification of making breaking changes to experimental features while developing mainstream features.  After all, we're not exactly a stable project per se.  Maybe once we achieve `1.0.0` is won't be as much of a concern.

> # Alternatives

Reviewer: What about experimental features that use RFCs but just for the subteam?

Reviewer 3: Personally I think experimental features in the lifecycle should typically (but not always) have RFCs b/c of the lifecycle's place in the overall architecture. However I don't RFCs should be necessary (most of the time) for experimental pack features.

Reviewer 3: But generally I think we should trust the judgement of maintainers when deciding whether an RFC is a good idea for an experimental feature.

Reviewer 2: Are you saying that the creator of an experimental feature wouldn't have to author an RFC, but the subteams would, as a process by which to accept/bless an experimental feature?

[alternatives]: #alternatives

## Plugins

> A plugin system for Pack and/or Lifecycle would allow experimental features to be implemented completely outside of the CNB project. For example, a plugin could be installed to Pack with a command like:

Reviewer 2: @jkutner is this suggesting we support experimental 3rd party plugins, or do you envision a separate CNB owned repo where these experimental plugins could live?

Author: Yea, I was thinking 3rd party (where the most likely 3rd parties would be Salesforce, Paketo, etc). But not a CNB owned repo.

Reviewer: I think a plugin system would be great independent of this RFC.

```

> $ pack install-plugin skele-gro

Reviewer 4: 👍  to the HP reference 🧹

```

Then, a user can execute new commands provided by the plugin:

```
$ pack grow-bones
```

Or use new flags on existing commands:

```
$ pack build --with-bone-growth
```

Such a plugin mechanism would require a well defined plugin-interface, and versioning of that interface. This is a significantly larger project that the current proposal. However, the current proposal does not prevent a plugin system from being introduced later on.

# Prior Art
[prior-art]: #prior-art

> - Rust's [Unstable features](https://doc.rust-lang.org/rustdoc/unstable-features.html)

Reviewer: FWIW, unstable features in rust (mostly) go through RFCs

- Java's [`-XX:+UnlockExperimentalVMOptions` flag](https://bugs.openjdk.java.net/browse/JDK-6618726)
- [oclif Plugins](https://oclif.io/docs/plugins)

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

> - How does an experimental features become a real feature (does it use the same flag/command minus the `--experimental` flag)?

Reviewer: Wouldn't it go through a RFC?

Author: i guess i don't mean the process, but the mechanics. Like, do we need to enforce that they are the same flag?

Reviewer: Why would it need to be the same flag? If an experimental feature "graduates" it could have changes etc. from the feedback/things we learn.

Reviewer 2: I wonder if we need an experimental flag?   I don't know how golang supports external plugins, but I can envision the command just being added to the existing list of commands, and the user would use it as any other command.   We could then mark experimental commands in the help menu or introduce an additional command to list all the experimental commands that have been installed (probably might want to do this anyway so users can quickly identify and remove any experimental commands).  

```
Available Commands:
  build                     Generate app image from source code
  grow-bones (experimental) Some custom experimental thingy
  rebase                    Rebase app image with latest run image
  inspect-image             Show information about a built image
  set-run-image-mirrors     Set mirrors to other repositories for a given run image
  set-default-builder       Set default builder used by other commands
```

# Meta
[meta]: #meta
- Name: Space and Process for Experimental Features
- Start Date: 2020-07-02
- Author(s): [@jkutner](https://github.com/jkutner)
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: N/A

# Summary
[summary]: #summary

This is a proposal for allowing experimental features without requiring a specification, core team review, or RFC.

# Motivation
[motivation]: #motivation

End-users and early adopters of CNB have real problems that need to be solved by the CNB project. Sometimes these problems prevent them from adopting CNB in their organizations. Our RFC process, specification guidelines, and governance structure provide a strong mechanism through which we can enact change and solve problems; but sometimes it's too heavy and inhibits experimentation.

> We need a way to introduce experiments before writing an RFC. Such a mechanism would allow us to deliver new capabilities faster, improve our means of collecting real end-user feedback, and improve the quality of our RFCs.

Reviewer 3: Doesn't `pack` already have experimental features? I think the UX around pack experimental features could maybe use some improvement, but they do exist.
